code/sonalyze/daemon/perform.go

```diff
@@ -0,0 +1,387 @@
+		default:
+			panic("Unexpected")
+		}
+	}
+}
+
+func httpPostHandler(
+	dc *DaemonCommand,
+	dataType string,
+	contentType string,
+) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		payload, userName, clusterName, ok := requestPreamble(dc, w, r, "POST", dc.postAuthenticator, "", contentType)
+		if !ok {
+			return
+		}
+
+		if dc.matchUserAndCluster && userName != "" && clusterName != userName {
+			w.WriteHeader(400)
+			fmt.Fprintf(w, "Upload not authorized")
+			if dc.Verbose {
+				statusWarningf("Upload not authorized")
+			}
+			return
+		}
+
+		verb := "add"
+		arguments := []string{
+			"--" + dataType,
+			"--data-path",
+			path.Join(dc.jobanalyzerDir, "data", clusterName),
+		}
+
+		stdout, ok := runSonalyze(dc, w, verb, arguments, payload)
+		if !ok {
+			return
+		}
+
+		w.WriteHeader(200)
+		fmt.Fprint(w, stdout)
+	}
+}
+
+func requestPreamble(
+	dc *DaemonCommand,
+	w http.ResponseWriter,
+	r *http.Request,
+	method string,
+	authenticator *auth.Authenticator,
+	realm string,
+	contentType string,
+) (payload []byte, userName, clusterName string, ok bool) {
+	if dc.Verbose {
+		// Header reveals auth info, don't put it into logs
+		statusInfof("Request from %s: %v", r.RemoteAddr, r.URL.String())
+	}
+
+	if !httpsrv.AssertMethod(w, r, method, dc.Verbose) {
+		return
+	}
+
+	authOk, userName := httpsrv.Authenticate(w, r, authenticator, realm, dc.Verbose)
+	if !authOk {
+		return
+	}
+
+	payload, havePayload := httpsrv.ReadPayload(w, r, dc.Verbose)
+	if !havePayload {
+		return
+	}
+
+	if contentType != "" {
+		if !httpsrv.AssertContentType(w, r, contentType, dc.Verbose) {
+			return
+		}
+	}
+
+	clusterValues, found := r.URL.Query()["cluster"]
+	if !found || len(clusterValues) != 1 || clusterValues[0] == "" {
+		w.WriteHeader(400)
+		fmt.Fprintf(w, "Bad parameters - missing or empty or repeated 'cluster'")
+		if dc.Verbose {
+			statusWarningf("Bad parameters - missing or empty or repeated 'cluster'")
+		}
+		return
+	}
+
+	clusterName = clusterValues[0]
+	if dc.aliasResolver != nil {
+		clusterName = dc.aliasResolver.Resolve(clusterName)
+	}
+
+	ok = true
+	return
+}
+
+func runSonalyze(dc *DaemonCommand, w http.ResponseWriter, verb string, arguments []string, input []byte) (stdout string, ok bool) {
+	cmdName := "<sonalyze>"
+
+	// Run the command and report the result
+
+	if dc.Verbose {
+		statusInfof(
+			"Command: %s %s",
+			path.Join(dc.jobanalyzerDir, cmdName),
+			verb+" "+strings.Join(arguments, " "),
+		)
+	}
+
+	anyCmd, _ := dc.cmdlineHandler.ParseVerb(cmdName, verb)
+	if anyCmd == nil {
+		errResponse(w, 400, fmt.Errorf("Bad verb in remote command: %s", verb), "", dc.Verbose)
+		return
+	}
+	fs := flag.NewFlagSet(cmdName, flag.ContinueOnError)
+	err := dc.cmdlineHandler.ParseArgs(verb, arguments, anyCmd, fs)
+	if err != nil {
+		errResponse(w, 400, err, "", dc.Verbose)
+		return
+	}
+	stop, err := dc.cmdlineHandler.StartCPUProfile(anyCmd.CpuProfileFile())
```

Comment:
It does not seem natural to start a CPU profile here?  It may also be tied into the problem in sonalyze.go where we pass a standard command line handler; a custom handler might just return an error code for a request for a cpu profile.